RInAppMessaging/Classes/Repositories/CampaignRepository.swift

```diff
@@ -70,29 +89,50 @@ internal class CampaignRepository: CampaignRepositoryType {
         let updatedCampaign = Campaign.updatedCampaign(campaign, asOptedOut: true)
         list[index] = updatedCampaign
         self.campaigns.set(value: list)
+        userDataCache.cacheCampaignData(list, userIdentifiers: preferenceRepository.getUserIdentifiers())
         return updatedCampaign
     }
 
     @discardableResult
-    func decrementImpressionsLeftInCampaign(_ campaign: Campaign) -> Campaign? {
-        let updatedCampaign = Campaign.updatedCampaign(campaign, withImpressionLeft: max(0, campaign.impressionsLeft - 1))
-        return updateCampaignInTheList(updatedCampaign) ? updatedCampaign : nil
+    func decrementImpressionsLeftInCampaign(id: String) -> Campaign? {
+        guard let campaign = findCampaign(withID: id) else {
+            return nil
+        }
+        return updateImpressionsLeftInCampaign(campaign, newValue: max(0, campaign.impressionsLeft - 1))
     }
 
     @discardableResult
-    func incrementImpressionsLeftInCampaign(_ campaign: Campaign) -> Campaign? {
-        let updatedCampaign = Campaign.updatedCampaign(campaign, withImpressionLeft: campaign.impressionsLeft + 1)
-        return updateCampaignInTheList(updatedCampaign) ? updatedCampaign : nil
+    func incrementImpressionsLeftInCampaign(id: String) -> Campaign? {
+        guard let campaign = findCampaign(withID: id) else {
+            return nil
+        }
+        return updateImpressionsLeftInCampaign(campaign, newValue: campaign.impressionsLeft + 1)
     }
 
-    private func updateCampaignInTheList(_ campaign: Campaign) -> Bool {
+    func loadCachedData() {
+        guard let cachedData = userDataCache.getUserData(identifiers: preferenceRepository.getUserIdentifiers())?.campaignData else {
+            return
+        }
+        campaigns.set(value: cachedData)
+    }
+
+    private func findCampaign(withID id: String) -> Campaign? {
+        let list = campaigns.get()
+        return list.first(where: { $0.id == id })
```

Comment:
Could be:
`private func findCampaign(withID id: String) -> Campaign? {
    campaigns.get().first(where: { $0.id == id })
}`